lib/irb/completion.rb

```diff
@@ -109,338 +63,31 @@ def self.retrieve_files_to_require_relative_from_current_dir
-          else
-            []
-          end
-        end
-
-      when /^(\$[^.]*)$/
-        # global var
-        gvar = $1
-        all_gvars = global_variables.collect{|m| m.to_s}
-
-        if doc_namespace
-          all_gvars.find{ |i| i == gvar }
-        else
-          all_gvars.grep(Regexp.new(Regexp.quote(gvar)))
-        end
-
-      when /^([^.:"].*)(\.|::)([^.]*)$/
-        # variable.func or func.func
-        receiver = $1
-        sep = $2
-        message = $3
-
-        gv = bind.eval_global_variables.collect{|m| m.to_s}.push("true", "false", "nil")
-        lv = bind.local_variables.collect{|m| m.to_s}
-        iv = bind.eval_instance_variables.collect{|m| m.to_s}
-        cv = bind.eval_class_constants.collect{|m| m.to_s}
-
-        if (gv | lv | iv | cv).include?(receiver) or /^[A-Z]/ =~ receiver && /\./ !~ receiver
-          # foo.func and foo is var. OR
-          # foo::func and foo is var. OR
-          # foo::Const and foo is var. OR
-          # Foo::Bar.func
-          begin
-            candidates = []
-            rec = eval(receiver, bind)
-            if sep == "::" and rec.kind_of?(Module)
-              candidates = rec.constants.collect{|m| m.to_s}
-            end
-            candidates |= rec.methods.collect{|m| m.to_s}
-          rescue Exception
-            candidates = []
-          end
-        else
-          # func1.func2
-          candidates = []
-        end
-
-        if doc_namespace
-          rec_class = rec.is_a?(Module) ? rec : rec.class
-          "#{rec_class.name}#{sep}#{candidates.find{ |i| i == message }}"
-        else
-          select_message(receiver, message, candidates, sep)
-        end
-
-      when /^\.([^.]*)$/
-        # unknown(maybe String)
-
-        receiver = ""
-        message = $1
-
-        candidates = String.instance_methods(true).collect{|m| m.to_s}
-
-        if doc_namespace
-          "String.#{candidates.find{ |i| i == message }}"
-        else
-          select_message(receiver, message, candidates.sort)
-        end
-
-      else
-        if doc_namespace
-          vars = (bind.local_variables | bind.eval_instance_variables).collect{|m| m.to_s}
-          perfect_match_var = vars.find{|m| m.to_s == input}
-          if perfect_match_var
-            eval("#{perfect_match_var}.class.name", bind)
-          else
-            candidates = (bind.eval_methods | bind.eval_private_methods | bind.local_variables | bind.eval_instance_variables | bind.eval_class_constants).collect{|m| m.to_s}
-            candidates |= ReservedWords
-            candidates.find{ |i| i == input }
-          end
-        else
-          candidates = (bind.eval_methods | bind.eval_private_methods | bind.local_variables | bind.eval_instance_variables | bind.eval_class_constants).collect{|m| m.to_s}
-          candidates |= ReservedWords
-          candidates.grep(/^#{Regexp.quote(input)}/).sort
-        end
-      end
+    def self.retrieve_completion_doc_namespace(target)
```

Comment:
Let's also take `bind` from argument.

Reply:
I change to pass bind to completor's constructor.